File: Python/Largest Rectangle in Histogram.py
"""
Given n non-negative integers representing the histogram's bar height where the width of each bar is 1, find the area of largest rectangle in the histogram.


Above is a histogram where width of each bar is 1, given height = [2,1,5,6,2,3].


The largest rectangle is shown in the shaded area, which has area = 10 unit.

For example,
Given height = [2,1,5,6,2,3],
return 10.
"""

#Time O(n^2)
class Solution:
    # @param height, a list of integer
    # @return an integer
    def largestRectangleArea(self, height):
        res = 0
        
        for i in reversed(xrange(len(height))):
            if i == len(height) - 1 or height[i] > height[i+1]:
                Min = 9223372036854775807
                
                for j in reversed(xrange(i+1)):
                    Min = min(Min,height[j])
                    res = max(res,(i+1-j)*Min)
        
        return res

class Solution:
    # @param height, a list of integer
    # @return an integer
    def largestRectangleArea(self, height):
        if len(height) == 0 or height is None:
            return 0
        
        res = 0
        index = [-1]
        
        for i in range(len(height)):
            while index[-1] > -1:
                if height[index[-1]] > height[i]:
                    top = index.pop()
                    res = max(res,height[top] * (i - 1 - index[-1]))
                else:
                    break
            index.append(i)
        
        while index[-1] != -1:
            top = index.pop()
            res = max(res,height[top] * (len(height) - 1 - index[-1]))
        return res

File: Java/Largest Rectangle in Histogram.java
/*
Given n non-negative integers representing the histogram's bar height where the width of each bar is 1, find the area of largest rectangle in the histogram.
Above is a histogram where width of each bar is 1, given height = [2,1,5,6,2,3].


The largest rectangle is shown in the shaded area, which has area = 10 unit.

For example,
Given height = [2,1,5,6,2,3],
return 10.
*/

public class Solution {
    public int largestRectangleArea(int[] height) {
        if(height == null || height.length == 0) return 0;
        
        int result = 0; 
        Stack<Integer> stack = new Stack<>();
        for(int i = 0 ; i <= height.length; i++){
            int cur = (i == height.length) ? -1 : height[i];
            while(!stack.empty() && cur <= height[stack.peek()]){
                int h = height[stack.pop()];
                int w = stack.empty() ? i : i - stack.peek() - 1;
                result = Math.max(result, h * w);
            }
            stack.push(i);
        }
        return result;
    }
}